Sema: Rewrite partial applications into closures

slavapestov opened the pull request:
Fixes rdar://21289579 / https://bugs.swift.org/browse/SR-75.

There's a big follow-up cleanup, which is removing curry thunks and assorted machinery from SILGen. I'll land that separately.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please benchmark

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please benchmark

akyrtzi commented on test/Index/roles.swift:
```diff
@@ -175,9 +175,9 @@ class AClass {
+// CHECK: [[@LINE-1]]:16 | instance-method/Swift | foo() | [[AClass_foo_USR]] | Ref,Call | rel: 0
 let _ = AClass(x: 1).foo
-// CHECK: [[@LINE-1]]:22 | instance-method/Swift | foo() | [[AClass_foo_USR]] | Ref | rel: 0
+// CHECK: [[@LINE-1]]:22 | instance-method/Swift | foo() | [[AClass_foo_USR]] | Ref,Call | rel: 0
```
Marking these as "call sites" is inaccurate, could you correct this?

slavapestov replied:
This test passes with no changes in the latest version of the commit.

akyrtzi commented on test/Index/roles.swift:
```diff
@@ -352,7 +352,7 @@ let otherInstance = AStruct(x: 1)
 
 _ = AStruct.init(x:)
 // CHECK: [[@LINE-1]]:18 | instance-property/Swift | x | [[AStruct_x_USR]] | Ref | rel: 0
-// CHECK: [[@LINE-2]]:5 | struct/Swift | AStruct | [[AStruct_USR]] | Ref | rel: 0
```
Is the reference to `AStruct` in this line not reported anymore?

CodaFi commented on lib/Sema/CSApply.cpp:
```diff
@@ -604,7 +606,10 @@ namespace {
+        assert(member->getDeclContext()->isTypeContext());
         return 2;
+      } else if (auto *elt = dyn_cast<EnumElementDecl>(member)) {
+        return elt->getParameterList() ? 2 : 1;
```
Can the brains of this function just be `ValueDecl::getNumCurryLevels()`?  The assertions are still good to keep.

slavapestov replied:
Yup! Thanks.

CodaFi commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -1509,16 +1488,35 @@ ConstraintSystem::getTypeOfMemberReference(
-  Type type;
+  Type type = openedType;
+
+  if (!outerDC->getSelfProtocolDecl()) {
```
This is like the third or fourth time this block appears.  Can we centralize this?  One of them is going to fall out of sync one day...

slavapestov replied:
I only found two, but I'll extract it into a new method on ValueDecl.

CodaFi reviewed (approved):
I've only got nits.  The approach looks fantastic.

futurejones commented:
Can one of the admins verify this patch?

tkremenek commented:
@swift-ci clean test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
I dusted this off and fixed all the merge conflicts. Still need to address Robert's review feedback and a couple of other nits.

slavapestov committed:
Sema: Rewrite partial applications into closures

When a method is called with fewer than two parameter lists,
transform it into a fully-applied call by wrapping it in a
closure.

Eg,

Foo.bar => { self in { args... self.bar(args...) } }
foo.bar => { self in { args... self.bar(args...) } }(self)

super.bar => { args... in super.bar(args...) }

With this change, SILGen only ever sees fully-applied calls,
which will allow ripping out some code.

This new way of doing curry thunks fixes a long-standing bug
where unbound references to protocol methods did not work.

This is because such a reference must open the existential
*inside* the closure, after 'self' has been applied, whereas
the old SILGen implementation of curry thunks really wanted
the type of the method reference to match the opened type of
the method.

A follow-up cleanup will remove the SILGen curry thunk
implementation.

Fixes rdar://21289579 and https://bugs.swift.org/browse/SR-75.

akyrtzi commented:
Just a reminder that the output changes in `test/Index/roles.swift` are technically regressions and it would be great to address them before merging 🙏

slavapestov commented:
@akyrtzi Yep, they're already addressed in the latest version!

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please benchmark

slavapestov commented:
@swift-ci Please benchmark

slavapestov committed:
Profile: Run tests through to -emit-ir to get better coverage

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov committed:
IDE: Fix indexing for Sema-built curry thunks

slavapestov committed:
IDE: Clean up traversal logic in SourceEntityWalker

Try to make this a bit more self-documenting.

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please benchmark

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please benchmark

slavapestov commented:
@swift-ci Please benchmark

slavapestov commented:
@swift-ci Please benchmark

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test Linux

benrimmington commented on CHANGELOG.md:
```diff
@@ -46,6 +46,19 @@ Swift Next
+
+* [SR-75][]:
+
+  Unapplied references to protocol methods methods are now supported. Previously this
```
`methods methods` ➡️ `methods`

```diff
-Unapplied references to protocol methods methods are now supported. Previously this
+Unapplied references to protocol methods are now supported. Previously this
```

slavapestov replied:
Thanks!

benrimmington commented on CHANGELOG.md:
```diff
@@ -46,6 +46,19 @@ Swift Next
+    func play(catToy: Toy)
+  }
+
+  let fn = Cat.play
```
[Require parameter names when referencing to functions](https://forums.swift.org/t/require-parameter-names-when-referencing-to-functions/27048)?

```diff
-let fn = Cat.play
+let fn = Cat.play(catToy:)
```

slavapestov replied:
Not actually required, but I agree it makes the example more clear.

davezarzycki commented:
Hi @slavapestov – Congratulations on landing this and fixing a long standing issue. That's awesome! Do you know of (or foresee) any case where a `DotSyntaxBaseIgnoredExpr` would have an implicit autoclosure on the RHS that is NOT the thunk logic created by this pull request? Thanks in advance – Dave

slavapestov commented:
@davezarzycki That's a good question. It may very well be that the answer is "no". Do you have a further simplification in mind?

davezarzycki commented:
I don’t have a simplification in mind although it does seem like it could be simplified.

FWIW – This change broke a research branch I’m working on and I wanted to know how paranoid my misc diagnostic should be after this pull request. I think I know now. Thanks
